[Merged by Bors] - feat(combinatorics): Define vertices of a Quiver.Path

or4nge19 opened the pull request:
This PR is linked to #27315 and #27325. It's part of PRing the formalization of core PF theorem here https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/Formalizing.20Perron-Frobenius/with/525516636

This adds:

Quiver.Path.vertices: A List V representing the sequence of vertices along a path.

vertices_comp: A proof that the vertices of a composed path p.comp q are p.vertices.dropLast ++ q.vertices. vertices_length: The length of the vertices list is p.length + 1. Lemmas connecting vertices to standard list operations like head?, getLast, and get. Lemmas for membership: start_mem_vertices and end_mem_vertices.

This is a prerequisite for defining path simplicity and formalizing cycle extraction, as used in the formalization of the Perron-Frobenius theorem

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

or4nge19 committed:
feat(combinatorics): Define vertices of a Quiver.Path

This PR is linked to #27315 and #27325. It's part of PRing the formalization of core PF theorem here https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/Formalizing.20Perron-Frobenius/with/525516636

This adds:

Quiver.Path.vertices: A List V representing the sequence of vertices along a path.
Quiver.Path.vertices_set: The Set of vertices visited.
Quiver.Path.vertices_finset: The Finset of vertices visited.
Quiver.Path.init_finset: The Finset of vertices excluding the final vertex of the path, needed for defining simple paths and cycles.

vertices_comp: A proof that the vertices of a composed path p.comp q are p.vertices.dropLast ++ q.vertices.
vertices_length: The length of the vertices list is p.length + 1.
Lemmas connecting vertices to standard list operations like head?, getLast, and get.
Lemmas for membership: start_mem_vertices and end_mem_vertices.

This is the prerequisite for defining path simplicity and formalizing cycle extraction, as used in the formalization of the Perron-Frobenius theorem

or4nge19 committed:
Update Vertices.lean

removed unused import

or4nge19 committed:
Update Vertices.lean

added end_map

or4nge19 committed:
Revert "Update Vertices.lean"

This reverts commit a706e88a95d037a65856bf20201532cc126b71f9.

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,262 @@
+  | b, _ => b
+
+@[simp]
+lemma end_cons {a b c : V} (p : Path a b) (e : b ⟶ c) : «end» (p.cons e) = c := rfl
```
```suggestion
lemma end_cons {a b c : V} (p : Path a b) (e : b ⟶ c) : (p.cons e).end = c := rfl
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,262 @@
+/-- The list of vertices in a path, including the start and end vertices. -/
+def vertices {a : V} : ∀ {b : V}, Path a b → List V
+  | _, nil => [a]
+  | _, cons p e => (p.vertices).concat («end» (p.cons e))
```
```suggestion
  | _, cons p e => (p.vertices).concat (p.cons e).end
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,262 @@
+  | _, nil => [a]
+  | _, cons p e => (p.vertices).concat («end» (p.cons e))
+
+/-- The set of vertices in a path. -/
+def vertices_set {a b : V} (p : Path a b) : Set V :=
+  {v | v ∈ p.vertices}
+
+/-- The finset of vertices in a path. -/
+def vertices_finset [DecidableEq V] {a b : V} (p : Path a b) : Finset V :=
+  p.vertices.toFinset
```
I'm not convinced we want these definitions since either we'll have to repeat a bunch of API for `List` (and especially `List.toFinset`) in this more restricted situation or else we'll constantly be unfolding them.

If you still feel we want them then I suggest that they should be `abbrev` and also note that by rule 4 of our [naming conventions](https://leanprover-community.github.io/contribute/naming.html#general-conventions) they should be `verticesSet` and `verticesFinset`.

or4nge19 replied:
I can turn them into [simp] lemmas and perhaps it works with less pollution

ocfnash replied:
Not having these extra `def`s is a big improvement.

I think that we will be able to avoid even the `simp` lemmas you highlight above, but we can return to this point after working on the rest of the PR if you wish.

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,262 @@
+    x ∈ p.init_finset ↔ x ∈ p.vertices.dropLast :=
+  List.mem_toFinset
+
+lemma vertices_nil (a : V) : (nil : Path a a).vertices = [a] := rfl
```
```suggestion
@[simp]
lemma vertices_nil (a : V) : (nil : Path a a).vertices = [a] := rfl
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,262 @@
+def vertices_finset [DecidableEq V] {a b : V} (p : Path a b) : Finset V :=
+  p.vertices.toFinset
+
+/-- The finset of vertices in a path, excluding the final vertex. -/
+def init_finset [DecidableEq V] {a b : V} (p : Path a b) : Finset V :=
+  p.vertices.dropLast.toFinset
```
I'm also a bit dubious about this definition and if we keep it we need a better name.

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,262 @@
+lemma vertices_cons {a b c : V} (p : Path a b) (e : b ⟶ c) :
+  (p.cons e).vertices = p.vertices.concat c := rfl
+
+-- Vertices of a path are always non-empty
+lemma vertices_nonempty' {V : Type*} [Quiver V] {a b : V} (p : Path a b) :
+    p.vertices.length > 0 := by
```
```suggestion
lemma length_vertices_pos {a b : V} (p : Path a b) :
    0 < p.vertices.length := by
```

ocfnash replied:
Also if we move this lemma below `vertices_length` then the proof can be just `simp`.

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,262 @@
+    simp only [List.concat_eq_append, List.length_append, List.length_cons, List.length_nil,
+      zero_add, gt_iff_lt, lt_add_iff_pos_left, add_pos_iff, ih, Nat.lt_one_iff, pos_of_gt, or_self]
+
+/-- The set of vertices in a path. -/
+def vertices_finset' {a b : V} (p : Path a b) : Set V :=
+  {v | v ∈ p.vertices}
+
```
This seems to be a duplicate of `vertices_set` above
```suggestion
```

ocfnash reviewed (commented):
Thanks for all this :)

I've left some initial review and will return for the rest as soon as you've had a chance to respond.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

or4nge19 committed:
Merge branch 'master' into mc_pf_quiver_vertices

or4nge19 committed:
Update Mathlib/Combinatorics/Quiver/Path/Vertices.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

or4nge19 committed:
Update Vertices.lean

turned the verticesSet, verticesFinset and initFinset into [simp] lemmas

or4nge19 committed:
fix lint

or4nge19 committed:
Merge branch 'master' into mc_pf_quiver_vertices

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

ocfnash committed:
Merge branch 'master' into mc_pf_quiver_vertices

ocfnash committed:
Fix `Mathlib.lean`

Achieved by running:
`lake exe mk_all`

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+lemma vertices_head? {a b : V} (p : Path a b) : p.vertices.head? = some a := by
+  induction p with
+  | nil => simp only [vertices_nil, List.head?_cons]
+  | cons p' e ih =>
+    simp only [vertices_cons]
+    have : ¬p'.vertices.isEmpty := by
+      simp_all only [List.isEmpty_iff]
+      apply Aesop.BuiltinRules.not_intro
+      intro a_1
+      simp_all only [List.head?_nil, reduceCtorEq]
+    simp
+    simp_all only [List.isEmpty_iff, Option.getD_some]
```
```suggestion
  | cons p' e ih => simp [ih]
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+    simp_all only [List.isEmpty_iff, Option.getD_some]
+
+@[simp]
+lemma length_toPath {a b : V} (e : a ⟶ b) : (e.toPath).length = 1 := rfl
```
```suggestion
lemma length_toPath {a b : V} (e : a ⟶ b) : e.toPath.length = 1 := rfl
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+lemma length_toPath {a b : V} (e : a ⟶ b) : (e.toPath).length = 1 := rfl
+
+@[simp]
+lemma vertices_getLast {a b : V} (p : Path a b) (h : p.vertices ≠ []) :
+  p.vertices.getLast h = b := by
```
We can use an autoParam for `h` here so that the user never has to supply it. This will also make the lemma fire in `simp` calls, which should be useful below.
```suggestion
lemma vertices_getLast {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_nonempty) :
    p.vertices.getLast h = b := by
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+  p.vertices.getLast h = b := by
+  induction p with
+  | nil => simp only [vertices_nil, List.getLast_singleton]
+  | cons p' e ih =>
+    simp only [vertices_cons]
+    rw [@List.getLast_concat']
```
```suggestion
  | cons p' e ih => simp
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+  (p.comp q).vertices = p.vertices.dropLast ++ q.vertices := by
+  induction q with
+  | nil =>
+    simp only [comp_nil, vertices_nil]
+    have h_nonempty : p.vertices.length > 0 := by simp
+    have h_ne_nil : p.vertices ≠ [] := List.ne_nil_of_length_pos h_nonempty
+    rw [← List.dropLast_append_getLast h_ne_nil, vertices_getLast p h_ne_nil]
+    simp_all only [gt_iff_lt, ne_eq, List.cons_ne_self, not_false_eq_true,
+      List.dropLast_append_of_ne_nil, List.dropLast_singleton, List.append_nil]
+  | cons q' e ih =>
+    simp only [comp_cons, vertices_cons, ih, List.concat_eq_append]
+    rw [List.append_assoc]
```
```suggestion
  | nil => simpa using (List.dropLast_append_getLast p.vertices_nonempty).symm
  | cons q' e ih => simp [ih]
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+
+lemma start_mem_vertices {a b : V} (p : Path a b) : a ∈ p.vertices := by
+  induction p with
+  | nil => simp only [vertices_nil, List.mem_cons, List.not_mem_nil, or_false]
+  | cons p' e ih =>
+    simp only [vertices_cons]
+    simp_all only [List.concat_eq_append, List.mem_append, List.mem_cons, List.not_mem_nil,
+      or_false, true_or]
```
```suggestion
  | nil => simp
  | cons p' e ih => simp [ih]
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+
+@[simp] lemma length_eq_zero_iff {a : V} (p : Path a a) :
+    p.length = 0 ↔ p = Path.nil := by
+  cases p
+  · simp only [length_nil]
+  · constructor
+    · intro h; cases h
+    · intro h; cases h
```
```suggestion
  cases p <;> tauto
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+    · intro h; cases h
+    · intro h; cases h
+
+/-- Given vertices lists from a path composition, the prefix path's vertices is a prefix of
+the full path's vertices -/
+lemma isPrefix_dropLast_of_comp_eq {V : Type*}
+    [Quiver V] {a b c : V} {p : Path a b} {p₁ : Path a c} {p₂ : Path c b}
+    (h : p.vertices = p₁.vertices.dropLast ++ p₂.vertices) :
+    p₁.vertices.dropLast.IsPrefix p.vertices := by
+  rw [h]
+  exact List.prefix_append p₁.vertices.dropLast p₂.vertices
+
```
I think this is too specialised / too thin a wrapper around `List.prefix_append` to be worth having.
```suggestion
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,236 @@
+
+/-- The head of the vertices list is the start vertex. -/
+@[simp]
+lemma vertices_head_eq_start {a b : V} (p : Path a b) :
+    p.vertices.head (vertices_nonempty p) = a := by
+  induction p with
+  | nil => simp only [vertices_nil, List.head_cons]
+  | cons p' _ ih =>
+    simp only [vertices_cons, List.concat_eq_append]
+    have : p'.vertices ≠ [] := vertices_nonempty p'
+    simp only [List.head_append_of_ne_nil this]
+    exact ih
```
```suggestion
lemma vertices_head_eq {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_nonempty) :
    p.vertices.head h = a := by
  induction p with
  | nil => simp only [vertices_nil, List.head_cons]
  | cons p' _ ih => simp [List.head_append_of_ne_nil (vertices_nonempty p'), ih]
```

ocfnash reviewed (commented):
I've left a bunch more suggestions (mostly just golfs / nitpicks). I think one more round of nitpicks and this will probably be ready :)

Note that you should remove the "awaiting-author" label when you're ready for review, otherwise your PR won't show up in our review queue (I only found this now because I went looking for it).

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

or4nge19 committed:
address comments

ok to all :) thanks so much

or4nge19 commented:
apologies for the delay in addressing these comments :) i have committed all of them, still will cross-check after CI is over. Thanks so much for your help. I'm working now on the next PRs towards PF

or4nge19 committed:
minor fix

remove unused string

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+  | _, nil => [a]
+  | _, cons p e => (p.vertices).concat (p.cons e).end
+
+@[simp]
+lemma mem_verticesSet_iff {a b : V} (p : Path a b) (v : V) :
+    v ∈ {v | v ∈ p.vertices} ↔ v ∈ p.vertices := Iff.rfl
+
+lemma mem_verticesFinset_iff [DecidableEq V] {a b : V} (p : Path a b) {x : V} :
+    x ∈ p.vertices.toFinset ↔ x ∈ p.vertices :=
+  List.mem_toFinset
+
+lemma mem_initFinset_iff [DecidableEq V] {a b : V} (p : Path a b) {x : V} :
+    x ∈ p.vertices.dropLast.toFinset ↔ x ∈ p.vertices.dropLast :=
+  List.mem_toFinset
+
```
Let's drop these:
```suggestion
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+@[simp]
+lemma vertices_cons {a b c : V} (p : Path a b) (e : b ⟶ c) :
+  (p.cons e).vertices = p.vertices.concat c := rfl
+
+lemma mem_verticesSet [DecidableEq V] {a b : V} (p : Path a b) (v : V) :
+    v ∈ p.vertices.toFinset ↔ v ∈ p.vertices := by simp
```
This also can go since it's just a specialisation of `List.mem_toFinset`:
```suggestion
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+  simp only [vertices_nil, List.mem_singleton, Set.ext_iff, Set.mem_singleton_iff]
+  exact fun x ↦ Set.mem_setOf
+
+@[simp]
+lemma verticesSet_cons {a b c : V} (p : Path a b) :
+  {v | v ∈ p.vertices ∨ v = c} = {v | v ∈ p.vertices} ∪ {c} := by
+  simp only [Set.union_singleton]
+  ext v
+  simp only [mem_verticesSet_iff, Set.mem_insert_iff]
+  rw [or_comm]; simp
+
```
This should also go:
```suggestion
```
After all, it's just a result about `Set`s really and is even provable via `aesop`. E.g.,
```lean
example {α : Type*} (s : Set α) (a : α) : {x | x ∈ s ∨ x = a} = s ∪ {a} := by aesop
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+  | nil => simp only [vertices_nil, List.head?_cons]
+  | cons p' e ih => simp [ih]
+
+@[simp]
+lemma length_toPath {a b : V} (e : a ⟶ b) : e.toPath.length = 1 := rfl
```
Please move this to the file `Combinatorics.Quiver.Path`, somewhere just below the definition of [Quiver.Path.length](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Combinatorics/Quiver/Path.html#Quiver.Path.length)

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+  | cons p' _ ih => simp [List.head_append_of_ne_nil (vertices_ne_nil p'), ih]
+
+/-- The last element of the vertices list is the end vertex. -/
+lemma vertices_getLast_eq_end {a b : V} (p : Path a b) :
+  p.vertices.getLast (vertices_ne_nil p) = b := by
```
Let's drop the `_end` since we directly reference `b` and also add an autoParam for the not-nil condition to make this lemma slightly more useful when rewriting:
```suggestion
lemma vertices_getLast_eq {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_ne_nil) :
  p.vertices.getLast h = b := by
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+  p.vertices.getLast (vertices_ne_nil p) = b := by
+  exact vertices_getLast p (vertices_ne_nil p)
+
+lemma end_prefix_eq_get_vertices {a b c : V} (p₁ : Path a c) (p₂ : Path c b) :
+    c = (p₁.comp p₂).vertices.get
+        ⟨p₁.length, by
+  simp only [vertices_comp, List.length_append, List.length_dropLast,
+    vertices_length, add_tsub_cancel_right, lt_add_iff_pos_right, add_pos_iff,
+    Nat.lt_one_iff, pos_of_gt, or_true]⟩ := by
+  simp only [List.get_eq_getElem, vertices_comp, List.length_dropLast, vertices_length,
+    add_tsub_cancel_right, le_refl, List.getElem_append_right, tsub_self, List.getElem_zero,
+    vertices_head_eq]
```
How about we flip the sides? We can also golf the proofs a bit and make this provable by `simp` if we first introduce the following lemma just below `vertices_head?` higher up:
```lean
@[simp]
lemma getElem_vertices_zero {a b : V} (p : Path a b) : p.vertices[0] = a := by
  induction p with
  | nil => simp
  | cons p' e ih => simp [ih]
```
so that this lemma can become
```suggestion
lemma vertices_comp_get_length_eq {a b c : V} (p₁ : Path a c) (p₂ : Path c b)
    (h : p₁.length < (p₁.comp p₂).vertices.length := by simp) :
    (p₁.comp p₂).vertices.get ⟨p₁.length, h⟩ = c := by
  simp
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+@[simp]
+lemma vertices_toPath {i j : V} (e : i ⟶ j) :
+    (e.toPath).vertices = [i, j] := by
+  change (Path.nil.cons e).vertices = [i, j]
+  simp [vertices_cons, vertices_nil];
```
```suggestion
  simp [Hom.toPath]
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+
+lemma vertices_toPath_tail {i j : V} (e : i ⟶ j) :
+    (e.toPath).vertices.tail = [j] := by
+  simp [vertices_toPath]
```
```suggestion
  simp
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -0,0 +1,198 @@
+/-- If a composition is `nil`, the left component must be `nil`
+    (proved via lengths, avoiding dependent pattern-matching). -/
+lemma nil_of_comp_eq_nil_left {a b : V} {p : Path a b} {q : Path b a}
+    (h : p.comp q = (Path.nil : Path a a)) : p.length = 0 := by
```
```suggestion
    (h : p.comp q = Path.nil) : p.length = 0 := by
```

ocfnash reviewed (approved):
Thanks for sticking with this.

I've left one final round of suggestions. Please apply these, and then feel free to merge!

bors d+

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

or4nge19 committed:
implements all comments, move length_toPath to quiver.Path

or4nge19 commented:
> > I'm not sure I can/have authorisation to remove the 'awaiting author' label as I see no 'x' button in the label window on the right. If ok in the meantime 'll try to ping you with request review when ready
> 
> It's certainly fine (and in fact quite useful) to ping me for review but are you sure you cannot remove the label? It should look like this:
> 
>  GH-labels.mov

I don't have the settings icon in those windows, it may be something unlocked for maintainers or reviewers :)

or4nge19 committed:
fix unused import

or4nge19 committed:
fix deprecated imports
